Firestore/core/src/firebase/firestore/remote/impl/reader.h

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_IMPL_READER_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_IMPL_READER_H_
+
+#include <pb.h>
+#include <pb_decode.h>
+
+#include <cstdint>
+#include <functional>
+#include <string>
+
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
+#include "Firestore/core/src/firebase/firestore/remote/impl/tag.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+namespace impl {
```

Comment:
Did we ever formally decide on whether or not we preferred `impl` or `internal`? When it's just an extra namespace in regular header in the package it hasn't mattered as much, but with this change it's materializing into the directory structure so it seems worth settling.

I mention it because Firebase C++ seems to favor `internal`.

Secondarily it seems like this is a bunch of stuff required for interacting with nanopb. When we implement the local serializer (which has types outside those required for the remote serializer) it's going to need access to these classes too.

Maybe it's worth pushing this up into `firebase/firestore/nanopb` instead of having classes in the local namespace referring to classes in `firebase/firestore/remote/internal`?

Reply:
> Did we ever formally decide on whether or not we preferred impl or internal? When it's just an extra namespace in regular header in the package it hasn't mattered as much, but with this change it's materializing into the directory structure so it seems worth settling.

> I mention it because Firebase C++ seems to favor internal.

I don't recall if we did or not; I just picked "impl". But if the rest of firebase uses internal, then I don't see any reason why we shouldn't. (But this is no longer relevant due to below.)

> Secondarily it seems like this is a bunch of stuff required for interacting with nanopb. When we implement the local serializer (which has types outside those required for the remote serializer) it's going to need access to these classes too.

> Maybe it's worth pushing this up into firebase/firestore/nanopb instead of having classes in the local namespace referring to classes in firebase/firestore/remote/internal?

Yeah, that's a good idea. Done.